src/connectors/configure-related-items/connectConfigureRelatedItems.ts

```diff
@@ -0,0 +1,157 @@
+import algoliasearchHelper, {
+  SearchParameters,
+  PlainSearchParameters,
+} from 'algoliasearch-helper';
+import { Unmounter, WidgetFactory, ResultHit } from '../../types';
+import { createDocumentationMessageGenerator, warning } from '../../lib/utils';
+import connectConfigure, {
+  ConfigureRenderer,
+  ConfigureConnectorParams,
+} from '../configure/connectConfigure';
+
+export type MatchingPatterns = {
+  [attribute: string]: {
+    /**
+     * The score of the optional filter.
+     *
+     * @see https://www.algolia.com/doc/guides/managing-results/rules/merchandising-and-promoting/in-depth/optional-filters/
+     */
+    score: number;
+  };
+};
+
+export interface ConfigureRelatedItemsConnectorParams {
+  /**
+   * The reference hit to extract the filters from.
+   */
+  hit: ResultHit;
+  /**
+   * The schema to create the optional filters.
+   * Each key represents an attribute from the hit.
+   */
+  matchingPatterns: MatchingPatterns;
+  /**
+   * Function called to transform the search parameters generated.
+   */
+  transformSearchParameters?(
+    searchParameters: SearchParameters
+  ): PlainSearchParameters;
+}
+
+export type ConfigureRelatedItemsWidgetFactory<
+  TConfigureRelatedItemsWidgetParams
+> = WidgetFactory<
+  ConfigureRelatedItemsConnectorParams & TConfigureRelatedItemsWidgetParams
+>;
+
+type ConfigureRelatedItemsConnector = <TConfigureRelatedItemsWidgetParams>(
+  render?: ConfigureRenderer<ConfigureConnectorParams>,
+  unmount?: Unmounter
+) => ConfigureRelatedItemsWidgetFactory<TConfigureRelatedItemsWidgetParams>;
+
+const withUsage = createDocumentationMessageGenerator({
+  name: 'configure-related-items',
+  connector: true,
+});
+
+function createOptionalFilter({
+  attributeName,
+  attributeValue,
+  attributeScore,
+}) {
+  return `${attributeName}:${attributeValue}<score=${attributeScore || 1}>`;
+}
+
+const connectConfigureRelatedItems: ConfigureRelatedItemsConnector = (
+  renderFn,
+  unmountFn
+) => {
+  warning(
+    false,
+    'ConfigureRelatedItems is an experimental widget that is subject to change in next minor versions.'
+  );
+
+  return widgetParams => {
+    const { hit, matchingPatterns, transformSearchParameters = x => x } =
+      widgetParams || ({} as typeof widgetParams);
+
+    if (!hit) {
```

Comment:
if these errors are done before the destructuring, the cast isn't needed I think.

Reply:
We'll need the cast anyway because we want to support passing an empty object (`{}`).

Reply:
isn't hit required?

Reply:
It is. Not sure to follow what you're suggesting. Can you elaborate what you have in mind?

Note that destructuring is also useful here to assign a default value to `transformSearchParameters`.

Reply:
If done after the ifs, we don’t need a default object anymore, since it’s always going to be a value. It’s more of a general pattern, so not too important to land with this PR